docs/tutorials/network/espnow.mdx

```diff
@@ -0,0 +1,257 @@
+
+<Note>
+
+When working with two devices, you will need to tell Jaguar which
+USB port to use.
+
+When using `jag monitor` you have to use the `--port` option to specify
+the port the ESP32 device is connected to. For example:
+
+``` shell
+jag monitor --port COM3          # Windows.
+jag monitor --port /dev/ttyUSB0  # Linux and macOS.
+```
+
+</Note>
+
+## Introduction
+
+ESP-NOW can be seen as a replacement for Wifi, where devices don't
+need to connect to an access point, but can communicate directly with
+each other. Its range is typically better than Wifi, but it has
+lower data rates.
+
+Devices can either communicate by broadcasting messages to all
+devices in range, or by sending messages to specific devices using
+their MAC address. In both cases, a key can be used to encrypt the
+messages.
+
+## Limitations
+
+Messages are at most 1470 bytes long. For broadcast messages there
+is no guarantee that messages are correctly delivered. For direct
+messages, the implementation ensures that the PHY layer has
+received the message, but there is no guarantee that the message
+is processed by the receiving device.
+
+ESP-NOW only supports 20 peers, although one of them can be the
+broadcast address, which represents all devices in range.
+
+By default only 7 keys can be used. This can be changed in the
+sdkconfig, but that requires building a custom envelope.
+
+In theory, ESP-NOW can run at the same times as Wifi, but Toit
+does not currently support this. If Wifi is enabled the following
+error will be raised when trying to create an ESP-NOW service:
+
+``` plain
+EXCEPTION error.
+ALREADY_IN_USE
+  0: espnow-create_            <sdk>/esp32/espnow.toit:336:3
+  ...
+```
+
+Since Jaguar automatically connects to Wifi when it starts up,
+we need to tell Jaguar to disable Wifi when running the programs.
+You can do this by `jag run` with `-Djag.wifi=false`. This disables
+Wifi while the given program is running. To make sure that the device is
+reachable, Jaguar sets a timeout of 10 seconds, after which the program
+is killed and Wifi is re-enabled. If you need more time, you can also
+pass the `-Djag.timeout=<duration>` option. For example, to run a
+program for 60 seconds, you can use `-Djag.timeout=1m`.
+
+A complete command line could look like this:
+
+``` shell
+jag run -Djag.wifi=false -Djag.timeout=1m sender.toit
+```
+
+## Broadcasting
+
+In this example we will set up one device to send broadcast messages
+and another device to receive them.
+
+### Sender
+
+Create a new file `sender.toit` and put the following code into it:
+
+```toit
+import esp32.espnow
+
+main:
+  service := espnow.Service
+  service.add-peer espnow.BROADCAST-ADDRESS
+
+  counter := 0
+  while true:
+    message := "hello $counter"
+    service.send message
+        --address=espnow.BROADCAST-ADDRESS
+    print "Sent datagram."
+    counter++
+    sleep --ms=1000
+```
+
+The program starts by creating a service, followed by adding the broadcast
```

Comment:
I still don't like mixing peers/devices and addresses like this.

Reply:
done.